File: music_kit_platform_interface/lib/model/catalog_song_response.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'artwork.dart';

part 'catalog_song_response.freezed.dart';
part 'catalog_song_response.g.dart';

@freezed
class CatalogSongResponse with _$CatalogSongResponse {
  const factory CatalogSongResponse({
    String? id,
    String? href,
    String? type,
    Song? attributes,
  }) = _CatalogSongResponse;

  factory CatalogSongResponse.fromJson(Map<String, dynamic> json) =>
      _$CatalogSongResponseFromJson(json);
}

enum AudioVariant {
  dolbyAtmos,
  dolbyAudio,
  lossless,
  highResolutionLossless,
  lossyStereo
}

@freezed
class Song with _$Song {
  const factory Song({
    int? discNumber,
    Artwork? artwork,
    String? name,
    String? releaseDate,
    String? composerName,
    List<PreviewAsset>? previews,
    bool? hasLyrics,
    int? trackNumber,
    String? url,
    List<AudioVariant>? audioVariants,
    String? isrc,
    PlayParams? playParams,
    bool? isAppleDigitalMaster,
    int? durationInMillis,
    String? contentRating,
    List<String>? genreNames,
    String? albumName,
    String? artistName,
  }) = _Song;

  factory Song.fromJson(Map<String, dynamic> json) => _$SongFromJson(json);
}

@freezed
class PreviewAsset with _$PreviewAsset {
  const factory PreviewAsset({
    String? url,
  }) = _PreviewAsset;

  factory PreviewAsset.fromJson(Map<String, dynamic> json) =>
      _$PreviewAssetFromJson(json);
}

@freezed
class PlayParams with _$PlayParams {
  const factory PlayParams({
    String? catalogId,
    String? id,
    String? kind,
    @JsonKey(name: 'musicKit_databaseID') String? musicKitDatabaseID,
    @JsonKey(name: 'musicKit_persistentID') String? musicKitPersistentID,
  }) = _PlayParams;

  factory PlayParams.fromJson(Map<String, dynamic> json) =>
      _$PlayParamsFromJson(json);
}

File: music_kit/android/src/main/kotlin/app/misi/music_kit/infrastructure/CatalogSongResponse.kt
package app.misi.music_kit.infrastructure

import kotlinx.serialization.*
import kotlinx.serialization.json.JsonNames

@Serializable
data class CatalogSongResponse (
    val data: List<CatalogSongResponseItem>? = null,
)

@Serializable
data class CatalogSongResponseItem (
    val id: String? = null,
    val type: String? = null,
    val href: String? = null,
    val attributes: Song? = null,
    val relationships: Relationships? = null
)

@Serializable
data class Song (
    val albumName: String? = null,
    val genreNames: List<String>? = null,
    val trackNumber: Long? = null,
    val releaseDate: String? = null,
    val durationInMillis: Long? = null,
    val isrc: String? = null,
    val artwork: Artwork? = null,
    val composerName: String? = null,
    val url: String? = null,
    val playParams: PlayParams? = null,
    val discNumber: Long? = null,
    val isAppleDigitalMaster: Boolean? = null,
    val hasLyrics: Boolean? = null,
    val name: String? = null,
    val previews: List<Preview>? = null,
    val artistName: String? = null
)

@Serializable
data class Artwork (
    val width: Long? = null,
    val height: Long? = null,
    val url: String? = null,
    val bgColor: String? = null,
    val textColor1: String? = null,
    val textColor2: String? = null,
    val textColor3: String? = null,
    val textColor4: String? = null
)

@Serializable
data class PlayParams (
    val id: String? = null,
    val kind: String? = null
)

@Serializable
data class Preview (
    val url: String? = null
)

@Serializable
data class Relationships (
    val albums: Albums? = null,
    val artists: Albums? = null
)

@Serializable
data class Albums (
    val href: String? = null,
    val data: List<ItemInfo>? = null
)

@Serializable
data class ItemInfo (
    val id: String? = null,
    val type: String? = null,
    val href: String? = null
)
